### ISSUE TITLE
transform_exclusive_scan fails sliently when passing __device__ only operator

### ISSUE THREAD

**Issue by cctry** [NVIDIA/thrust#1740]
I am using transform_exclusive_scan to work on an array of struct. I passed an operator as defined below. 
The code can be compiled but fails silently, and the result checking does not run. 

Adding `__host__` makes the code run correctly but `ncu` cannot profile any kernel. The compiled SASS only contains `cub11EmptyKernel` with no meaningful operation. 

The code is compiled with `nvcc test.cu -O3 -arch=sm_75` using CUDA 11.6.
```
#include <bits/stdc++.h>
#include <thrust/transform_scan.h>

struct Matrix {
  int rows, cols;
};

int main() {
  int N = 1 << 24;
  Matrix *matrices;
  cudaMallocManaged(&matrices, N * sizeof(Matrix));
  for (int i = 0; i < N; i++) {
    matrices[i].rows = rand() % 5 + 1;
    matrices[i].cols = rand() % 5 + 1;
  }
  int *offset;
  cudaMallocManaged(&offset, N * sizeof(int));
  // thrust implementation
  struct OP : thrust::unary_function<Matrix, int> {
    // NOTE: Add __host__ for correct results
    __device__ float operator()(Matrix m) { return m.rows * m.cols; }
  };
  thrust::transform_exclusive_scan(matrices, matrices + N, offset, OP(), 0, thrust::plus<int>());
  // reference implementation
  std::vector<int> offset_ref(N);
  offset_ref[0] = 0;
  for (int i = 1; i < N; i++) {
    offset_ref[i] =
        offset_ref[i - 1] + matrices[i - 1].rows * matrices[i - 1].cols;
  }
  // check results
  cudaDeviceSynchronize();
  for (int i = 0; i < N; i++) {
    if (offset[i] != offset_ref[i]) {
      printf("Error: offset[%d] = %d, expected %d\n", i, offset[i], offset_ref[i]);
      return 1;
    }
  }
  printf("Success\n");
  return 0;
}
```

**Comment by jrhemstad** [comment]
TL;DR: You need to use a `thrust::device` execution policy to get the algorithm to run on the GPU. 
```
  thrust::transform_exclusive_scan(thrust::device, matrices, matrices + N, offset, OP(), 0, thrust::plus<int>());
```

In general, when you invoke a Thrust algorithm without an execution policy it will attempt to deduce the execution policy based on the inputs. For instance, if you were to do:
```
thrust::device_vector<int> ints(...);
thrust::sort(ints.begin(), ints.end());
```

This would be able to deduce that the algorithm should execute on the device because it recognizes the iterator type of `device_vector` and statically dispatches to the device execution path.

However, in your situation the input to the `transform_exclusive_scan` are just raw pointers. There is nothing in the type system for Thrust to infer the execution space and defaults to executing on the host.

Therefore, you need to explicitly request execution on the device by specifying the `thrust::device` execution policy.

**Comment by cctry** [comment]
Thank you very much for the explanation.
